bin/docker-setup

```diff
@@ -1,10 +1,26 @@
 #!/usr/bin/env sh
 
+set -e
+
 echo "== Building Docker image (this may take a while) =="
 docker-compose build
 
+echo "== Ensuring Elasticsearch data directory is owned by the correct uid =="
+docker-compose run elasticsearch chown elasticsearch:elasticsearch /usr/share/elasticsearch/data
+
 echo "== Setting up database =="
-docker-compose run web rails db:setup db:migrate search:setup data_updates:run
+on_db_setup_failed() {
+  echo "Database setup failed with exit code $1." >&2
+  echo "" >&2
+  echo "Check the Docker logs using:" >&2
+  echo "    docker-compose logs" >&2
+  echo "" >&2
+  echo "Once you fixed the error, try running the script again." >&2
+  echo "To also delete all data, run this before retrying:" >&2
+  echo "    docker-compose down -v && sudo rm -rf docker_data" >&2
+  exit $1
+}
+docker-compose run web sh -c 'yarn wait-on -t 30000 http-get://elasticsearch:9200/ && rails search:setup db:setup db:migrate data_updates:run' || on_db_setup_failed $?
```

Comment:
First off, hi Thai, fancy seeing you here 😄 and thanks for the help! 🙏  

I'm a bit reluctant to add the `wait-on` package for this, as Docker Compose supports something similar natively ([healthcheck documentation](https://docs.docker.com/compose/compose-file/#/healthcheck#healthcheck). Containers that depend on Elasticsearch running, can declare the dependency like this:

```yaml
    depends_on:
      elasticsearch:
        condition: service_healthy
```

For the Elasticsearch container you can then define a health check like this:

```yaml
    healthcheck:
        test: ["CMD", "curl", "-f", "http://localhost:3000"]
        interval: 30s
        timeout: 30s
        retries: 3
```

See [here](https://docs.docker.com/engine/reference/builder/#healthcheck) for more informationr regarding interval, timeout and retries.

Reply:
@citizen428 Hey Michael! Thanks for your review, I just learned about healthcheck support just now. Never knew it existed before (I depend on npm for most things nowadays :joy:)

@maestromac But it works on my machine™! Guess we did not get what Docker promised here… 😅 Jokings aside, I was able to replicate this when trying to run the build on GitHub Actions. I will use healthcheck as @citizen428 suggested instead. Thanks for your review.

Reply:
@citizen428 I checked, but unfortunately [`condition` is no longer supported](https://docs.docker.com/compose/compose-file/#depends_on)…

> Version 3 no longer supports the condition form of `depends_on`.

[docker-compose’s docs](https://docs.docker.com/compose/startup-order/) suggests users to “Use a tool such as wait-for-it, dockerize, or sh-compatible wait-for”. So I am bringing back `wait-on` for now. Instead of adding it as project dependency, I made the DB setup install `wait-on` and wait for it ad-hoc instead.

Reply:
@maestromac 

I get it now… This happens because the `web` container has this in `docker-compose.yml`:

```
    volumes:
      - .:/usr/src/app
```

Now, `node_module` exists _inside_ the container, but not inside the host. That is why you get the command not found error. Now, why did I not get that error? I ran `yarn add wait-on` outside, so my local machine has `node_modules`.

This is fixed by adding another entry. It will cause `node_modules` to go into a separate volume with the initial contents copied from inside the container. (Thanks [StackOverflow](https://stackoverflow.com/a/32785014/559913)!)

```diff
     volumes:
       - .:/usr/src/app
+      - /usr/src/app/node_modules
```

Reply:
@dtinth Ah, sorry, I missed that they removed `depends_on`. The new solution seems good.

> But it works on my machine™! Guess we did not get what Docker promised here… 😅 

This is a depressingly common occurence for me. 🙄